test/parseedn-test-data.el

```diff
@@ -243,6 +243,36 @@
                                              (:form . "123")
                                              (:value . 123)))))))))
 
+       "prefixed-map"
+       (a-list
+        :source "#:foo.bar{:baz 1 :other.ns.prefix/qux 2}"
+        :edn (list (a-hash-table :foo.bar/baz 1 :other.ns.prefix/qux 2))
+        :ast '((:node-type . :root)
```

Comment:
You can also assert that the `:ast` of `"#:foo.bar{:baz}"` is equal to that of `{:foo.bar/baz}"`

Reply:
Well, that is not the case. `parseclj` produces different a different `:ast` for each case. Because it parses the map namespace (what `parseclj` calles the "map prefix") as a separate element of the AST. That is, it doesn't perform the `#:foo.bar{:baz 1}` -> `{:foo.bar/baz 1}` transformation as part of the parsing and AST generation. It produces two AST elements, one for the map prefix, and a second one for the map itself (note: I'm using `parseclj-parse-clojure` in the examples to make them self-contained, but the generated AST is exactly the same when parsing the examples as EDN instead as of Clojure):


```elisp
*** Welcome to IELM ***  Type (describe-mode) or press C-h m for help.
ELISP> (require 'parseclj)
parseclj
ELISP> (parseclj-parse-clojure "#::foo.bar{:baz 1}")
((:node-type . :root)
 (:position . 1)
 (:children
   ((:map-prefix
      (:token-type . :map-prefix)
      (:form . "#::foo.bar")
      (:pos . 1))
    (:node-type . :map)
    (:position . 11)
    (:children
      ((:node-type . :keyword)
       (:position . 12)
       (:form . ":baz")
       (:value . :baz))
      ((:node-type . :number)
       (:position . 17)
       (:form . "1")
       (:value . 1))))))

ELISP> 
```
You can see that the AST for `{:foo.bar/baz 1}` doesn't have the `:map-prefix` element at all:

```elisp
ELISP> (parseclj-parse-clojure "{:foo.bar/baz 1}")
((:node-type . :root)
 (:position . 1)
 (:children
   ((:node-type . :map)
    (:position . 1)
    (:children
      ((:node-type . :keyword)
       (:position . 2)
       (:form . ":foo.bar/baz")
       (:value . :foo.bar/baz))
      ((:node-type . :number)
       (:position . 15)
       (:form . "1")
       (:value . 1))))))

ELISP>
```

That is why we need to do the "assembling" (transformation) of the pieces in `parseedn-reduce-branch` (which is 95% of  what this patch does :smile:).

Reply:
Got it!

Still, would it be possible to assert a resemblance? e.g. `(= x (dissoc y :foo))`

What we gain is a higher-level test.

Otherwise, it can be easy for the `:ast`s to become sort of a blob without a particular meaning. In other words, a test that asserts that `(= x <large object>)` doesn't necessarily assert a specific property, so later, if the test changes, it's hard to see if it's still valid.

Reply:
First of all, thanks a lot for your feedback!

Unfortunately, I don't I understand what you mean by "assert a resemblance" and "gain a higher-level test". I don't know if you are referring to the `:ast` value when you talk about those two things, or the `:edn` and `:source` values.

I am not an expert in the `parseedn` package testing strategy, But from what I can understand from `define-parseedn-read-tests` and `define-parseedn-roundtrip-tests` (in `parsedn-test.el`), the `:ast` key defined in all the data definitions present in `parseedn-test-data.el`, that key is not used at all. So I would say that (at least at the moment), that key value is more informational that anything else.

Also, `parseedn-read` never sees the full value of the `:ast` as defined in `parseedn-test-data.el`. In fact, it doesn't know about it at all. The only functions that see anything related to the `:ast` value are `parseedn-reduce-leaf` and `parseedn-reduce-branch`. But they are called back from `parseclj` when needed, to "reduce" (transform the current AST node definition into the associated Emacs Lisp data type). In the former case to reduce the individual nodes of the `:ast` tree that are considered "leaf tokens" by `parseclj`. In the latter to reduce the sub-tree of the `:ast` tree that is considered a "branch" by `parseclj` (e.g., maps, lists, vectors, etc).

So could you please clarify what you'd like to test?

Reply:
I'll try this out locally, and if I find a meaningful test that can be added, will just post it here

Reply:
Merged!

Reply:
As you say, `parseedn-read never sees the full value of the :ast as defined in parseedn-test-data.el`.

So, let's remove all `:ast` values. They very likely are remnants of when parseedn and parseclj were the same library.

I have verified that tests keep passing.

If you want a patch:

    curl https://gist.githubusercontent.com/vemv/bbd7da50ffa8e82b685d905dfc276f4e/raw/43f411ac4b944d5e61643926b01b0cb09dfdd901/gistfile1.txt | git apply

Reply:
All `:ast` keys removed.